woocommerce/woocommerce
# Add logging and an admin notice for Legacy REST API usages

### Changes proposed in this Pull Request:

In order to help users determine how the Legacy REST API is used on its site in the context of [its removal in WooCommerce 9.0](https://developer.woo.com/2023/10/03/the-legacy-rest-api-will-move-to-a-dedicated-extension-in-woocommerce-9-0/), this pull request adds a helper tool that introduces two mechanisms:

1. Logging (in log files with a dedicated name) usages of the Legacy REST API, including API version, request route, user agent (when available) and date.
2. Displaying an admin notice with summary information about the requests received.

![image](https://github.com/woocommerce/woocommerce/assets/937723/b5071888-6376-4a93-8bc4-1ffa635e4231)

Internally, a transient named `wc_legacy_rest_api_usages` (with a lifetime of two weeks) is kept with the date of the first access, the total accesses count, and the different user agents registered.

The notice will not show, and logging won't happen, if any of these is true:

* The separate Legacy REST API extension is installed and active.
* The Legacy REST API is disabled.
* The transient with usages data is not available.

Additionally, it's possible to deactivate the writing of the log entry by having the `woocommerce_log_legacy_rest_api_usages` filter return `false`.

### How to test the changes in this Pull Request:

To test the changes in this pull request you need to perform requests to at least one endpoint of the Legacy REST API using Postman, `curl` or your favorite HTTP request tool. For example: `http://localhost/wc-api/v3/orders`.

1. Install [the Legacy REST API plugin](https://github.com/woocommerce/woocommerce-legacy-rest-api) but don't activate it. If you have it already installed and active, deactivate it.

2. Enable the Legacy REST API (in WooCommerce - Settings - Advanced - Legacy API) if it's not active.

2. Perform two Legacy REST API requests specifying different user agents, and then one more without user agent header.

3. Go to the logs page, verify that the notice has appeared and displays correct information (it should say that it has recognized threee accesses from two different known user agents) and that the log file has been created with the appropriate entries (the log file name starts with `legacy_rest_api_usages`).

4. Check the transient file contents with `wp transient get 'wc_legacy_rest_api_usages'`, then delete it with `wp transient delete wc_legacy_rest_api_usages` to simulate the transient expiration.

5. Reload any admin page. See how the notice has disappeared.

6. Perform yet another Legacy REST API request and reload any admin page, see how the notice is there again but its data has been reset (first date is current date and access count is 1).

7. Verify that disabling the Legacy REST API in settings makes the notice disappear, and enabling it again makes the notice appear again.

8. Activate the Legacy REST API plugin and reload the admin page, the notice has now disappeared. Verify also that Legacy REST API usages are no longer logged.

9. Deactivate the Legacy REST API plugin and reload the admin page, the notice will appear again and logging of new Legacy REST API usages will resume.

10. Dismiss the notice by clicking its "dismiss" link, it will no longer appear regardless of new Legacy REST API usages or extension activation or deactivation. If you want to undo the dismissal run this: `wp user meta delete <your user id> "dismissed_legacy_api_usages_detected_notice"`

11. Verify that Legacy REST API usages stop being logged (even without the extension active) when you add the following code: `add_filter('woocommerce_log_legacy_rest_api_usages', '__return_false');`


### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
